src/connector/src/source/google_pubsub/source/reader.rs

```diff
@@ -0,0 +1,125 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::time::Duration;
+
+use anyhow::{anyhow, ensure, Context, Ok, Result};
+use async_trait::async_trait;
+use chrono::{TimeZone, Utc};
+use futures_async_stream::try_stream;
+use google_cloud_pubsub::client::Client;
+use google_cloud_pubsub::subscription::{SeekTo, Subscription};
+use risingwave_common::try_match_expand;
+
+use super::TaggedReceivedMessage;
+use crate::source::google_pubsub::PubsubProperties;
+use crate::source::{
+    BoxSourceStream, Column, ConnectorState, SourceMessage, SplitId, SplitImpl, SplitReader,
+};
+
+const PUBSUB_MAX_FETCH_MESSAGES: usize = 1024;
+
+pub struct PubsubSplitReader {
+    subscription: Subscription,
+    split_id: SplitId,
+}
+
+impl PubsubSplitReader {
+    #[try_stream(boxed, ok = Vec<SourceMessage>, error = anyhow::Error)]
+    pub async fn into_stream(self) {
+        loop {
+            let raw_chunk = self
+                .subscription
+                .pull(PUBSUB_MAX_FETCH_MESSAGES as i32, None, None)
+                .await?;
+
+            // TODO: handle errors conditionally
+
+            // Sleep if we get an empty batch -- this should generally not happen
+            // since subscription.pull claims to block until at least a single message is available.
+            // But pull seems to time out at some point a return with no messages, so we need to see
+            // ? if that's somehow adjustable or we can skip sleeping and hand it off to pull again
+            if raw_chunk.is_empty() {
+                tokio::time::sleep(Duration::from_millis(200)).await;
+                continue;
+            }
+
+            let mut chunk: Vec<SourceMessage> = Vec::with_capacity(raw_chunk.len());
+            let mut ack_ids: Vec<String> = Vec::with_capacity(raw_chunk.len());
+
+            for message in raw_chunk {
+                ack_ids.push(message.ack_id().into());
+                chunk.push(SourceMessage::from(TaggedReceivedMessage(
+                    self.split_id.clone(),
+                    message,
+                )));
+            }
+
+            self.subscription
+                .ack(ack_ids)
+                .await
+                .map_err(|e| anyhow!(e))
+                .context("failed to ack pubsub messages")?;
+
+            yield chunk;
```

Comment:
please refer to https://github.com/risingwavelabs/risingwave/blob/4e5619c962d5649afbf2479be1eacd17c19b46a6/src/connector/src/source/kafka/source/reader.rs#L109-L118

yield a single message rather than a chunk

Reply:
Isn't the kafka reader also yielding a chunk currently?

Reply:
Oh sorry, my fault.